shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -124,19 +165,59 @@ public void setMessageHandler(
       @Nullable TaskQueue taskQueue) {
     if (handler == null) {
       Log.v(TAG, "Removing handler for channel '" + channel + "'");
-      messageHandlers.remove(channel);
-    } else {
-      DartMessengerTaskQueue dartMessengerTaskQueue = null;
-      if (taskQueue != null) {
-        dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
-        if (dartMessengerTaskQueue == null) {
-          throw new IllegalArgumentException(
-              "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
-        }
+      synchronized (handlersLock) {
+        messageHandlers.remove(channel);
+      }
+      return;
+    }
+    DartMessengerTaskQueue dartMessengerTaskQueue = null;
+    if (taskQueue != null) {
+      dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
+      if (dartMessengerTaskQueue == null) {
+        throw new IllegalArgumentException(
+            "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
       }
-      Log.v(TAG, "Setting handler for channel '" + channel + "'");
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
     }
+    runDelayedTasksForChannel(channel);
+  }
+
+  /**
+   * Runs the tasks that handle messages received from Dart for the provided channel name.
+   *
+   * <p>The channel may not have associated tasks if it was registered prior to reciving the first
+   * message from Dart.
+   *
+   * @param channel The channel name.
+   */
+  public void runDelayedTasksForChannel(@NonNull String channel) {
+    LinkedList<DelayedMessageInfo> list;
+    synchronized (handlersLock) {
+      if (!delayedTaskDispatcher.contains(channel)) {
+        return;
+      }
+      list = (LinkedList) delayedTaskDispatcher.get(channel);
+      delayedTaskDispatcher.remove(channel);
+    }
+    while (!list.isEmpty()) {
+      final DelayedMessageInfo info = list.poll();
+      dispatchMessageToQueue(
+          channel, messageHandlers.get(channel), info.message, info.replyId, info.messageData);
+    }
+  }
+
+  /**
+   * Enables the ability to queue tasks when messages are received from Dart.
+   *
+   * <p>This is useful when there are pending channel handler registrations. For example, Dart may
+   * be initialized concurrently, and prior to the registration of the channel handlers. This
+   * implies that Dart may start sending messages while plugins are being registered.
+   */
+  public void enableDelayedTaskQueue() {
```

Comment:
I'd prefer a different name to avoid reusing the term `TaskQueue`

Suggestion: `enableBufferingIncomingMessages`?

If this is renamed, the `canDelayTasks` flag name should also be changed to match